diff --git a/st.c b/st.c
index 64bc208..a7b4827 100644
--- a/st.c
+++ b/st.c
@@ -246,6 +246,7 @@ typedef struct {
 	Line *alt;    /* alternate screen */
 	Line hist[histsize]; /* history buffer */
 	int histi;    /* history index */
+	int histlen;  /* history length */
 	int scr;      /* scroll back */
 	int *dirty;  /* dirtyness of lines */
 	XftGlyphFontSpec *specbuf; /* font spec buffer used for rendering */
@@ -340,6 +341,8 @@ typedef struct {
 /* function definitions used in config.h */
 static void clipcopy(const Arg *);
 static void clippaste(const Arg *);
+static void kscrollend(const Arg *);
+static void kscrollbeg(const Arg *);
 static void kscrolldown(const Arg *);
 static void kscrollup(const Arg *);
 static void numlock(const Arg *);
@@ -1533,9 +1536,9 @@ ttywrite(const char *s, size_t n)
 	fd_set wfd, rfd;
 	ssize_t r;
 	size_t lim = 256;
-	Arg arg = (Arg) { .i = term.scr };
+	Arg arg = (Arg) { 0 };
 
-	kscrolldown(&arg);
+	kscrollend(&arg);
 
 	/*
 	 * Remember that we are using a pty, which might be a modem line.
@@ -1738,12 +1741,32 @@ tswapscreen(void)
 }
 
 void
+kscrollend(const Arg* a)
+{
+	int n = term.scr;
+	term.scr = 0;
+	selscroll(0, -n);
+	tfulldirt();
+}
+
+void
+kscrollbeg(const Arg* a)
+{
+	int n = term.histlen - term.scr;
+	if(n <= 0)
+		return;
+	term.scr += n;
+	selscroll(0, term.scr);
+	tfulldirt();
+}
+
+void
 kscrolldown(const Arg* a)
 {
 	int n = a->i;
 
 	if (n < 0)
-		n = term.row + n;
+		n = term.row + n + 1;
 
 	if (n > term.scr)
 		n = term.scr;
@@ -1761,9 +1784,10 @@ kscrollup(const Arg* a)
 	int n = a->i;
 
 	if (n < 0)
-		n = term.row + n;
+		n = term.row + n + 1;
+	n = MIN(term.histlen - term.scr, n);
 
-	if (term.scr <= histsize - n) {
+	if (n > 0) {
 		term.scr += n;
 		selscroll(0, n);
 		tfulldirt();
@@ -1783,6 +1807,9 @@ tscrolldown(int orig, int n, int copyhist)
 		temp = term.hist[term.histi];
 		term.hist[term.histi] = term.line[term.bot];
 		term.line[term.bot] = temp;
+		if(term.histlen > 0) {
+			--term.histlen;
+		}
 	}
 
 	tsetdirt(orig, term.bot-n);
@@ -1810,6 +1837,8 @@ tscrollup(int orig, int n, int copyhist)
 		temp = term.hist[term.histi];
 		term.hist[term.histi] = term.line[orig];
 		term.line[orig] = temp;
+		if(term.histlen < histsize)
+			++term.histlen;
 	}
 
 	tclearregion(0, orig, term.col-1, orig+n-1);
