diff --git a/config.def.h b/config.def.h
index 0d83e00..90e7544 100644
--- a/config.def.h
+++ b/config.def.h
@@ -186,6 +186,8 @@ Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_I,           iso14755,       {.i =  0} },
 	{ ShiftMask,            XK_Page_Up,     kscrollup,      {.i = -1} },
 	{ ShiftMask,            XK_Page_Down,   kscrolldown,    {.i = -1} },
+	{ ShiftMask,            XK_Home,        kscrollbeg,     {      0} },
+	{ ShiftMask,            XK_End,         kscrollend,     {      0} },
 };
 
 /*
diff --git a/st.c b/st.c
index d33eb5b..97fd3bd 100644
--- a/st.c
+++ b/st.c
@@ -128,6 +128,8 @@ typedef struct {
 /* function definitions used in config.h */
 static void clipcopy(const Arg *);
 static void clippaste(const Arg *);
+static void kscrollend(const Arg *);
+static void kscrollbeg(const Arg *);
 static void kscrolldown(const Arg *);
 static void kscrollup(const Arg *);
 static void numlock(const Arg *);
@@ -859,9 +861,9 @@ ttywrite(const char *s, size_t n)
 	fd_set wfd, rfd;
 	ssize_t r;
 	size_t lim = 256;
-	Arg arg = (Arg) { .i = term.scr };
+	Arg arg = (Arg) { 0 };
 
-	kscrolldown(&arg);
+	kscrollend(&arg);
 
 	/*
 	 * Remember that we are using a pty, which might be a modem line.
@@ -1064,6 +1066,28 @@ tswapscreen(void)
 }
 
 void
+kscrollend(const Arg* a)
+{
+	int n = term.scr;
+	if(n <= 0)
+		return;
+	term.scr = 0;
+	selscroll(0, -n);
+	tfulldirt();
+}
+
+void
+kscrollbeg(const Arg* a)
+{
+	int n = term.histlen - term.scr;
+	if(n <= 0)
+		return;
+	term.scr += n;
+	selscroll(0, n);
+	tfulldirt();
+}
+
+void
 kscrolldown(const Arg* a)
 {
 	int n = a->i;
@@ -1088,8 +1112,9 @@ kscrollup(const Arg* a)
 
 	if (n < 0)
 		n = term.row + n;
+	n = MIN(term.histlen - term.scr, n);
 
-	if (term.scr <= HISTSIZE-n) {
+	if (n > 0) {
 		term.scr += n;
 		selscroll(0, n);
 		tfulldirt();
@@ -1109,6 +1134,9 @@ tscrolldown(int orig, int n, int copyhist)
 		temp = term.hist[term.histi];
 		term.hist[term.histi] = term.line[term.bot];
 		term.line[term.bot] = temp;
+		if(term.histlen > 0) {
+			--term.histlen;
+		}
 	}
 
 	tsetdirt(orig, term.bot-n);
@@ -1136,6 +1164,8 @@ tscrollup(int orig, int n, int copyhist)
 		temp = term.hist[term.histi];
 		term.hist[term.histi] = term.line[orig];
 		term.line[orig] = temp;
+		if(term.histlen < histsize)
+			++term.histlen;
 	}
 
 	tclearregion(0, orig, term.col-1, orig+n-1);
diff --git a/st.h b/st.h
index ca90c31..df1673a 100644
--- a/st.h
+++ b/st.h
@@ -120,6 +120,7 @@ typedef struct {
 	Line *alt;    /* alternate screen */
 	Line hist[HISTSIZE]; /* history buffer */
 	int histi;    /* history index */
+	int histlen;  /* history length */
 	int scr;      /* scroll back */
 	int *dirty;  /* dirtyness of lines */
 	GlyphFontSpec *specbuf; /* font spec buffer used for rendering */
